awscli/customizations/logs/tail.py

```diff
@@ -294,20 +295,31 @@ def _filter_log_events(self, filter_logs_events_kwargs):
             return
 
     def _do_filter_log_events(self, filter_logs_events_kwargs):
-        last_event_ids = []
+        event_ids_per_timestamp = defaultdict(set)
         while True:
             response = self._client.filter_log_events(
                 **filter_logs_events_kwargs)
             for event in response['events']:
                 # For the case where we've hit the last page, we will be
-                # reusing the most recent nextToken to continue polling. This
-                # means it is possible that duplicate log events from the last
-                # page are returned back which we do not want to yield again.
+                # reusing the newest timestamp of the received events to keep polling.
+                # This means it is possible that duplicate log events with same timestamp
+                # are returned back which we do not want to yield again.
                 # We only want to yield log events that we have not seen.
-                if event['eventId'] not in last_event_ids:
+                if event['eventId'] not in event_ids_per_timestamp[event['timestamp']]:
+                    event_ids_per_timestamp[event['timestamp']].add(event['eventId'])
                     yield event
-            last_event_ids = [event['eventId'] for event in response['events']]
+            if event_ids_per_timestamp.keys():
```

Comment:
For dictionaries, we should be fine doing `if event_ids_pert_timestamp`. Specifically, the check is already based on the existence of keys for dictionaries.